### PR TITLE
pg_upgrade: Replace List with RBTree for preassigned Oids

### PR DESCRIPTION
**Pull request by danielgustafsson**
pg_upgrade can potentially preassign quite a large number of Oids when upgrading, so the List would be a bottleneck during lookups of Oids allocated to non-preassigned objects. This replaces the list with a Red-Black tree to achieve O(log n) lookups at the cost of slower inserts.

This diff isn't necessarily a win, since the List can do O(1) inserts, and I haven't benchmarked it really since I don't have an interesting dataset/schema handy. I had this sitting in my local tree, but since I'm not actively working on upgrades at the moment I wanted to submit it, if nothing else as a discussion piece.

@jchampio we had some discussions earlier about the scalability of the List implementation, do you guys have a pipeline where you can compare whats in HEAD with this?

### CODE REVIEW

File: src/backend/catalog/oid_dispatch.c [hunk 1]
```diff
@@ -735,6 +742,42 @@ GetPreassignedOidForType(Oid namespaceOid, const char *typname,
...
+
+static void
+rbtree_free(RBNode *node, void *arg)
+{
+	pfree(node);
+}
+
+/*
+ * The RB Tree combiner function will be called when a new node has the same
+ * key as an existing node (when rbtree_alloc() returns zero). For this
```
[the remark this answers is not in the document]
Comment (danielgustafsson) [reply src/backend/catalog/oid_dispatch.c]: Yes of course, will fix.

File: src/backend/catalog/oid_dispatch.c [hunk 2]
```diff
@@ -735,6 +742,42 @@ GetPreassignedOidForType(Oid namespaceOid, const char *typname,
...
+ * particular usecase the only value we have is the key, so make it a no-op.
+ */
+static void
+rbtree_combine(RBNode *existing __attribute__((unused)), const RBNode *new __attribute__((unused)), void *arg)
```
[the remark this answers is not in the document]
Comment (danielgustafsson) [reply src/backend/catalog/oid_dispatch.c]: Good point, thanks!

**Comment by danielgustafsson** [comment]
Pushed a rebase with the review comments addressed.

**Comment by danielgustafsson** [comment]
@jchampio have you guys had a chance to look at this patch yet? Just curious really.

**Comment by jchampio** [comment]
@danielgustafsson Not yet. It's queued up in Tracker and we actually just had a chat about it during backlog grooming today -- once the partition-constraint stuff finally lands, we'll get started on the performance testing.

**Comment by danielgustafsson** [comment]
> ..once the partition-constraint stuff finally lands, we'll get started on the performance testing.

No worries, thanks for the update!

**Comment by berlin-ab** [comment]
@jchampio Is it time to revisit this one yet?

**Comment by Eulerizeit** [comment]
On Tue, Mar 19, 2019 at 9:01 PM Adam Berlin <notifications@github.com>
wrote:

> @jchampio <https://github.com/jchampio> Is it time to revisit this one
> yet?
>

This is going to be ours

-- Rob

**Comment by danielgustafsson** [comment]
I'm inclined to say that we should merge this into master and not stress too much about it. The patch is likely to be a net negative on small schemas, but that's not really the usecase we're optimizing for.

**Comment by jchampio** [comment]
@berlin-ab Probably, yeah. When I wrote
> once the partition-constraint stuff finally lands

I hadn't expected that to be another two months in the making...

@danielgustafsson Merging into master seems reasonable, especially now that we've got a 6X branch.

**Comment by danielgustafsson** [comment]
Looking at randomly picked CI runs without this patch, and comparing with the latest runtime of this patch, it consistently performs ~15-20 seconds better than the linked list version. Pushed to master but won't backpatch this (at least not until it's been beaten around a bit).

**Pull request closed without merging**